creator-node/src/services/stateMachineManager/monitoring/StateMonitoringQueue.js

```diff
@@ -0,0 +1,254 @@
+        )}. Queuing another job...`
+      )
+      if (result?.jobFailed) {
+        this.enqueueJobAfterSuccess(queue, job, result)
+      } else {
+        this.enqueueJobAfterFailure(queue, job)
+      }
+    })
+    queue.on('failed', (job, err) => {
+      this.logError(
+        `Queue Job Failed - ID ${job?.id} - Error ${err}. Queuing another job...`
+      )
+      this.enqueueJobAfterFailure(queue, job)
+    })
+  }
+
+  /**
+   * Enqueues a job that processes the next slice of users after the slice
+   * that the previous job sucessfully processed.
+   * @param queue the StateMonitoringQueue to enqueue a job
+   * @param successfulJobData the jobData of the previous job that succeeded
+   * @param successfulJobResult the result of the previous job that succeeded
+   */
+  enqueueJobAfterSuccess(queue, successfulJobData, successfulJobResult) {
+    const {
+      data: { discoveryNodeEndpoint, moduloBase, currentModuloSlice }
+    } = successfulJobData
+    const { lastProcessedUserId } = successfulJobResult
+
+    queue.add({
+      startTime: Date.now(),
+      lastProcessedUserId,
+      discoveryNodeEndpoint,
+      moduloBase,
+      currentModuloSlice: (currentModuloSlice + 1) % moduloBase
+    })
+  }
+
+  /**
+   * Enqueues a job that picks up where the previous failed job left off.
+   * @param queue the StateMonitoringQueue to enqueue a job
+   * @param failedJob the jobData for the previous job that failed
+   */
+  enqueueJobAfterFailure(queue, failedJob) {
+    const {
+      data: {
+        lastProcessedUserId,
+        discoveryNodeEndpoint,
+        moduloBase,
+        currentModuloSlice
+      }
+    } = failedJob
+
+    queue.add({
+      startTime: Date.now(),
+      lastProcessedUserId,
+      discoveryNodeEndpoint,
+      moduloBase,
+      currentModuloSlice: (currentModuloSlice + 1) % moduloBase
+    })
+  }
+
+  /**
+   * Registers the logic that gets executed to process each new job from the queue.
+   * @param {*} queue the StateMonitoringQueue to consume jobs from
+   */
+  registerQueueConsumer(queue) {
+    // Initialize queue job processor (aka consumer)
+    queue.process(1 /** concurrency */, async (job) => {
+      const {
+        id: jobId,
+        data: {
+          lastProcessedUserId,
+          discoveryNodeEndpoint,
+          moduloBase,
+          currentModuloSlice
+        }
+      } = job
+
+      try {
+        this.log(`New job details: jobId=${jobId}, job=${JSON.stringify(job)}`)
+      } catch (e) {
+        this.logError(`Failed to log details for jobId=${jobId}: ${e}`)
+      }
+
+      // Default results of this job will be passed to the next job, so default to failure
+      let result = {
+        lastProcessedUserId,
+        jobFailed: true,
+        moduloBase,
+        currentModuloSlice
+      }
+      try {
+        // TODO: Wire up metrics
+        // await redis.set('stateMachineQueueLatestJobStart', Date.now())
+        result = await processStateMonitoringJob(
```

Comment:
nitty - since this is the StateMonitoringQueue, we can prob just call this processJob
idc tho

Reply:
synced about this. will get Vicky's input too but probably instead change util to consistently have the names in each file